pkg/loadbalancer/experimental/services_test.go

```diff
@@ -0,0 +1,411 @@
+// SPDX-License-Identifier: Apache-2.0
+// Copyright Authors of Cilium
+
+package experimental_test
+
+import (
+	"context"
+	"encoding/binary"
+	"encoding/json"
+	"log/slog"
+	"os"
+	"strings"
+	"testing"
+
+	"github.com/cilium/hive/cell"
+	"github.com/cilium/hive/hivetest"
+	"github.com/cilium/statedb"
+	"github.com/cilium/statedb/reconciler"
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
+
+	"github.com/cilium/cilium/pkg/cidr"
+	"github.com/cilium/cilium/pkg/clustermesh/types"
+	"github.com/cilium/cilium/pkg/hive"
+	"github.com/cilium/cilium/pkg/loadbalancer"
+	"github.com/cilium/cilium/pkg/loadbalancer/experimental"
+	"github.com/cilium/cilium/pkg/source"
+)
+
+type servicesParams struct {
+	cell.In
+
+	DB       *statedb.DB
+	Services *experimental.Services
+
+	FrontendTable statedb.Table[*experimental.Frontend]
+	BackendTable  statedb.Table[*experimental.Backend]
+}
+
+func servicesFixture(t testing.TB) (p servicesParams) {
+	log := hivetest.Logger(t, hivetest.LogLevel(slog.LevelError))
+
+	h := hive.New(
+		experimental.ServicesCell,
+
+		cell.Invoke(func(p_ servicesParams) { p = p_ }),
+	)
+
+	hive.AddConfigOverride(h, func(cfg *experimental.Config) {
+		cfg.EnableNewServices = true
+	})
+
+	require.NoError(t, h.Start(log, context.TODO()))
+	t.Cleanup(func() {
+		h.Stop(log, context.TODO())
+	})
+	return p
+}
+
+func intToAddr(i int) types.AddrCluster {
+	var addr [4]byte
+	binary.BigEndian.PutUint32(addr[:], 0x0100_0000+uint32(i))
+	addrCluster, _ := types.AddrClusterFromIP(addr[:])
+	return addrCluster
+}
+
+func TestServices_Service_UpsertDelete(t *testing.T) {
+	p := servicesFixture(t)
+	name := loadbalancer.ServiceName{Namespace: "test", Name: "test1"}
+	addrCluster := intToAddr(1)
+	frontend := *loadbalancer.NewL3n4Addr(loadbalancer.TCP, addrCluster, 12345, loadbalancer.ScopeExternal)
+
+	// Add a dump of the state if the test fails. Note that we abort
+	// the delete write transactions so they're not visible via this.
+	t.Cleanup(func() {
+		if t.Failed() {
+			p.Services.DebugDump(p.DB.ReadTxn(), os.Stdout)
+		}
+	})
+
+	// UpsertService
+	{
+		wtxn := p.Services.WriteTxn()
+		created, err := p.Services.UpsertFrontend(
+			wtxn,
+			&experimental.FrontendParams{
+				Name:    name,
+				Address: frontend,
+				Type:    loadbalancer.SVCTypeClusterIP,
+				Source:  source.Kubernetes,
+			},
+		)
+		require.True(t, created, "Service created")
+		require.NoError(t, err, "UpsertService")
+		wtxn.Commit()
+	}
+
+	// Lookup
+	{
+		txn := p.DB.ReadTxn()
```

Comment:
Please correct me if I'm missing something here. It seems more consistent to expose `ReadTxn` in `Services` (similarly to how we expose `func (s *Services) Frontends() statedb.Table[*Frontend]` for read-only access) so that we can write here `txn := p.Services.ReadTxn()` analogous to `wtxn := p.Services.WriteTxn()` in line 82 and 119.

Reply:
Agreed.